homeassistant/components/pi_hole/__init__.py

```diff
@@ -16,95 +16,172 @@
 from homeassistant.helpers import config_validation as cv
 from homeassistant.helpers.aiohttp_client import async_get_clientsession
 from homeassistant.helpers.discovery import async_load_platform
-from homeassistant.util import Throttle
+from homeassistant.util import Throttle, slugify
 
 from .const import (
     DOMAIN,
     CONF_LOCATION,
-    DEFAULT_HOST,
     DEFAULT_LOCATION,
     DEFAULT_NAME,
     DEFAULT_SSL,
     DEFAULT_VERIFY_SSL,
     MIN_TIME_BETWEEN_UPDATES,
     SERVICE_DISABLE,
     SERVICE_DISABLE_ATTR_DURATION,
+    SERVICE_DISABLE_ATTR_NAME,
     SERVICE_ENABLE,
+    SERVICE_ENABLE_ATTR_NAME,
 )
 
+
+def ensure_unique_names(config):
+    """Ensure that each configuration dict contains a unique `name` value."""
+    names = {}
+    for conf in config:
+        if conf[CONF_NAME] not in names:
+            names[conf[CONF_NAME]] = conf[CONF_HOST]
+        else:
+            raise vol.Invalid(
+                "Duplicate name '{}' for '{}' (already in use by '{}'). Each configured Pi-hole must have a unique name.".format(
+                    conf[CONF_NAME], conf[CONF_HOST], names[conf[CONF_NAME]]
+                )
+            )
+    return config
+
+
 LOGGER = logging.getLogger(__name__)
 
-CONFIG_SCHEMA = vol.Schema(
+PI_HOLE_SCHEMA = vol.Schema(
     {
-        DOMAIN: vol.Schema(
-            {
-                vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
-                vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-                vol.Optional(CONF_API_KEY): cv.string,
-                vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
-                vol.Optional(CONF_LOCATION, default=DEFAULT_LOCATION): cv.string,
-                vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
-            }
-        )
-    },
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+        vol.Optional(CONF_LOCATION, default=DEFAULT_LOCATION): cv.string,
+        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.All([PI_HOLE_SCHEMA], ensure_unique_names))},
```

Comment:
Add `cv.ensure_list` first inside `vol.All`. Then it's not necessarily a breaking change.